### PR TITLE
fnm: new port

### PR DESCRIPTION
**Pull request by manojkarthick**
#### Description

fnm is a fast and simple Node.js version manager, built in Rust

###### Type(s)

- [ ] bugfix
- [x] enhancement
- [ ] security fix

###### Tested on

macOS 11.4 20F71 x86_64
Xcode 12.5 12E262

###### Verification 
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [ ] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL? 
- [x] checked your Portfile with `port lint`?
- [ ] tried existing tests with `sudo port test`?
- [x] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?

### CODE REVIEW

**Comment by herbygillot** [comment]
I suppose if `fnm` manages and installs its own versions of Node, then that's somewhat at odds with MacPorts' ports of the same.

**Comment by manojkarthick** [comment]
I see, I'll close this PR then. Thanks!

**Comment by herbygillot** [comment]
I didn't mean that this should be unilaterally closed, just as something which should be discussed.

**Review by harens — COMMENTED** [review]
Thanks again for another great contribution @manojkarthick! A few notes from me below.

- It might be useful to add some notes about the [shell setup](https://github.com/Schniz/fnm#shell-setup).
- fnm offers [tests](https://github.com/Schniz/fnm#running-tests) (haven't tested this code yet):

```tcl
test.run            yes
test.cmd            ${cargo.bin} test
test.args           --target [cargo.rust_platform]
```

- Based on the [homebrew formula](https://github.com/Homebrew/homebrew-core/blob/master/Formula/fnm.rb), we can also add shell completion:

```tcl
post-destroot {
    set fnm_bin ${destroot}/${prefix}/bin/${name}

    # bash completion
    set bash_completion_dir ${destroot}${prefix}/etc/bash_completion.d
    xinstall -d ${bash_completion_dir}
    system "${fnm_bin} completions --shell==bash > ${bash_completion_dir}/${name}"

    # zsh completion
    set zsh_completion_dir ${destroot}${prefix}/share/zsh/site-functions
    xinstall -d ${zsh_completion_dir}
    system "${fnm_bin} completions --shell==zsh > ${zsh_completion_dir}/_${name}"

    # fish completion
    set fish_completion_dir ${destroot}${prefix}/share/fish/completions
    xinstall -d ${fish_completion_dir}
    system "${fnm_bin} completions --shell==fish > ${fish_completion_dir}/${name}.fish"
}
```

**Comment by harens** [comment]
> I suppose if fnm manages and installs its own versions of Node, then that's somewhat at odds with MacPorts' ports of the same.

@herbygillot that's a good catch, but I don't think this is a problem. fnm installs its node versions into `~/.fnm` where it works fine. Also, Homebrew offer node/npm but they haven't made any special modifications to their [fnm formula](https://github.com/Homebrew/homebrew-core/blob/master/Formula/fnm.rb).

However, this does mean it installs outside of the standard prefix. We either need some way of changing this or we need to note that this happens, similar to the [npm ports](https://github.com/macports/macports-ports/blob/94a298cfaa349a1e1afcc182effb015ba51db5d7/devel/npm6/Portfile#L93-L98).

**Comment by manojkarthick** [comment]
apologies @herbygillot , I figured `fnm` was redundant since `port select` seems to provide similar functionality. Happy to work on this if you think there's value to still include this port.

@harens I can update the portfile and add notes on the location. As for the other option you mentioned,`fnm` seems to provide an option for changing the install dir. If we go down the route of changing this, what would be a more appropriate location?

File: devel/fnm/Portfile [hunk 1]
```diff
@@ -0,0 +1,269 @@
+# -*- coding: utf-8; mode: tcl; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- vim:fenc=utf-8:ft=tcl:et:sw=4:ts=4:sts=4
+
+PortSystem          1.0
+PortGroup           github 1.0
+PortGroup           cargo 1.0
+
+github.setup        Schniz fnm 1.26.0 v
+name                fnm
+revision            0
```
[the remark this answers is not in the document]
Comment (manojkarthick) [reply devel/fnm/Portfile:9]: Thanks @harens , I will update the Portfile to remove name and add the tests/completion scripts.

**Comment by harens** [comment]
> @harens I can update the portfile and add notes on the location. As for the other option you mentioned,fnm seems to provide an option for changing the install dir. If we go down the route of changing this, what would be a more appropriate location?

My first thought was in `${prefix}/libexec/.fnm`, but on second thoughts, my worry is that it won't have write permission to install new node versions. @herbygillot do you know anyway around this? Otherwise we can just add it to the notes.

**Comment by herbygillot** [comment]
> > @harens I can update the portfile and add notes on the location. As for the other option you mentioned,fnm seems to provide an option for changing the install dir. If we go down the route of changing this, what would be a more appropriate location?
> 
> My first thought was in `${prefix}/libexec/.fnm`, but on second thoughts, my worry is that it won't have write permission to install new node versions. @herbygillot do you know anyway around this? Otherwise we can just add it to the notes.

I'd say just let it do what it does by default and let it install into `$HOME/.fnm`.

**Comment by harens** [comment]
> I'd say just let it do what it does by default and let it install into $HOME/.fnm.

That's fair enough. We can just note this in some [npm-style](https://github.com/macports/macports-ports/blob/94a298cfaa349a1e1afcc182effb015ba51db5d7/devel/npm6/Portfile#L93-L98) notes, and then after sorting out the other things I mentioned, I think we're done. Thanks @herbygillot for your suggestion, and @manojkarthick for your work on this.

**Comment by herbygillot** [comment]
> That's fair enough. We can just note this in some [npm-style](https://github.com/macports/macports-ports/blob/94a298cfaa349a1e1afcc182effb015ba51db5d7/devel/npm6/Portfile#L93-L98) notes, and then after sorting out the other things I mentioned, I think we're done. Thanks @herbygillot for your suggestion, and @manojkarthick for your work on this.

So I think someone who is installing `fnm` already has the reasonable expectation that it will install Node versions to their home directory as they use it.  I suppose it can't hurt to remind them that these will stick around via port notes when they install, uninstall or upgrade `fnm`, but that may not be entirely necessary.

**Comment by harens** [comment]
> So I think someone who is installing fnm already has the reasonable expectation that it will install Node versions to their home directory as they use it. I suppose it can't hurt to remind them that these will stick around via port notes when they install, uninstall or upgrade fnm, but that may not be entirely necessary.

I think it's more so that they know it won't be uninstalled with the port and that they'll have to do it manually (since it wasn't present during destroot). However, you're right in that they might already be aware of this.

**Comment by herbygillot** [comment]
> I think it's more so that they know it won't be uninstalled with the port and that they'll have to do it manually

Makes sense - thinking about it again, it's probably good to make an explicit note of this.

**Comment by harens** [comment]
@manojkarthick apologies for the delay in getting this sorted. If you're still willing to update this port (no rush of course), that would be greatly appreciated.

Otherwise, no worries :) I'm more than happy to make the changes myself. Thanks for creating this new port @manojkarthick.

**Comment by manojkarthick** [comment]
@harens It's been a hectic few weeks, I'll work on this in the next couple of days. Thanks!

**Review by harens — APPROVED** [review]
LGTM. Thanks @manojkarthick for your contribution! 🎉

**Pull request merged**